docs/SPIR-V.rst

```diff
@@ -0,0 +1,194 @@
+
+Library
+-------
+
+On the library side, this means introducing a new ``ASTFrontendAction`` and a SPIR-V module builder.  The new frontend action will traverse the AST and call the SPIR-V module builder to construct SPIR-V words. These code should be placed at ``tools/clang/lib/SPIRV`` and packed into one library (or multiple libraries in the future).
+
+Detailed design will be revised to accommodate more and more HLSL features. At the moment, we have::
+
+                EmitSPIRVAction
+                     |
+                     | creates
+                     V
+                SPIRVEmitter
+                     |
+                     | contains
+                     |
+       +-------------+------------+
+       |                          |
+       |                          |
+       V         references       V
+  SPIRVContext <------------ ModuleBuilder
+                                  |
+                                  | contains
+                                  V
+                              InstBuilder
+                                  |
+                                  | depends on
+                                  V
+                             WordConsumer
+
+``SPIRVEmitter``
+  The derived ``ASTConsumer`` which acts on various frontend AST nodes by calling corresponding ``ModuleBuilder`` methods to build SPIR-V modules gradually.
+``ModuleBuilder``
+  Exposes API for constructing SPIR-V modules. Interally it has structured representation of SPIR-V modules, functions, basic blocks as well as various SPIR-V specific structs like entry points, debug names, and so on.
+``SPIRVContext``
+  Responsible for <result-id> allocation and maintaining the lifetime of objects allocated to represent types, decorations, and others. It is used in conjunction with ``ModuleBuilder``.
+``InstBuilder``
+  The low-level interface for generating SPIR-V words for various SPIR-V instructions. All SPIR-V instructions are eventually serialized via ``InstBuilder``.
+``WordConsumer``
+  The consumer of generated SPIR-V words.
+
+Command-line tool
+-----------------
+
+On the command-line tool side, this means introducing a new binary, ``hlsl2spirv`` to wrap around the library functionality.
+
+But as the initial scaffolding step, a new option, ``-spirv``, will be added into ``dxc`` for invoking the new SPIR-V codegen action.
+
+Build system
+------------
+
+SPIR-V codegen functionality will structured as an optional feature in DirectXShaderCompiler. Two new CMake options will be introduced to control the configuring and building SPIR-V codegen:
+
+``ENABLE_SPIRV_CODEGEN``
+  If turned on, enables the SPIR-V codegen functionality.
+``SPIRV_BUILD_TESTS``
+  If turned on, enables building of SPIR-V related tests. This option will also implicitly turns on ``ENABLE_SPIRV_CODEGEN``.
+
+For the ``hctbuild`` script, the corresponding switches for the above two options are ``-spirv`` and ``-spirvtest``.
+
```

Comment:
also add that running `hcttest spirv` will run all DirectXShaderCompiler tests as well as SPIR-V related tests.